### PR TITLE
executor:adjust 'show create table' for  mysql compatibility

### PR DESCRIPTION
**Pull request by xiekeyi98**
show create table should not print the charset of the column
  if it is the same with table charset( same with MySQL)

### What problem does this PR solve? 

If the charset of column is the same with table charset , we should better not show the carset of the column.
(In order to be consistent with MySQL.)

### What is changed and how it works?

Change `func (e *ShowExec) fetchShowCreateTable() error{}` , to determine if the column charset and table charset are equal.

### Check List 

Tests 

 - Unit test

### Before
```sql
mysql> show create table ptest\G
*************************** 1. row ***************************
       Table: ptest
Create Table: CREATE TABLE `ptest` (
  `a` int(11) NOT NULL,
  `b` double NOT NULL DEFAULT '2.0',
  `c` varchar(10) CHARSET utf8mb4 COLLATE utf8mb4_bin NOT NULL,
  `d` time DEFAULT NULL,
  `e` timestamp NULL DEFAULT NULL,
  `f` timestamp NULL DEFAULT NULL,
  PRIMARY KEY (`a`),
  UNIQUE KEY `d` (`d`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
1 row in set (0.00 sec)
```

### After
```sql
mysql> show create table ptest\G
ERROR 2006 (HY000): MySQL server has gone away
No connection. Trying to reconnect...
Connection id:    1
Current database: test

*************************** 1. row ***************************
       Table: ptest
Create Table: CREATE TABLE `ptest` (
  `a` int(11) NOT NULL,
  `b` double NOT NULL DEFAULT '2.0',
  `c` varchar(10) NOT NULL,
  `d` time DEFAULT NULL,
  `e` timestamp NULL DEFAULT NULL,
  `f` timestamp NULL DEFAULT NULL,
  PRIMARY KEY (`a`),
  UNIQUE KEY `d` (`d`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
1 row in set (0.00 sec)
```

### CODE REVIEW

**Comment by AndrewDi** [comment]
@xiekeyi98 For your reference in my test, if we create column with specific charset, even column's charset is the same as table's charset, `SHOW CREATE TABLE` will show `CHARACTER SET utf8mb4 COLLATE utf8mb4_0900_ai_ci` part. 
```mysql
mysql> use test
Database changed
mysql> create table t (a int,b varchar(20));
Query OK, 0 rows affected (0.02 sec)

mysql> show create table t\G
*************************** 1. row ***************************
       Table: t
Create Table: CREATE TABLE `t` (
  `a` int(11) DEFAULT NULL,
  `b` varchar(20) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
1 row in set (0.00 sec)

mysql> create table t1 (a int,b varchar(20) charset utf8mb4);
Query OK, 0 rows affected (0.02 sec)

mysql> show create table t1\G
*************************** 1. row ***************************
       Table: t1
Create Table: CREATE TABLE `t1` (
  `a` int(11) DEFAULT NULL,
  `b` varchar(20) CHARACTER SET utf8mb4 COLLATE utf8mb4_0900_ai_ci DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
1 row in set (0.00 sec)

mysql> select version()\G
*************************** 1. row ***************************
version(): 8.0.15
1 row in set (0.00 sec)

```

**Comment by xiekeyi98** [comment]
It seems I make something wrong(sorry for the lack of testing).

The following is the MYSQL impl:
```c++
    if (field->has_charset()) {
      /*
        For string types dump charset name only if field charset is same as
        table charset or was explicitly assigned.
      */
      if (field->charset() != share->table_charset ||
          column_has_explicit_collation) {
        packet->append(STRING_WITH_LEN(" CHARACTER SET "));
        packet->append(field->charset()->csname);
      }
      /*
        For string types dump collation name only if
        collation is not primary for the given charset
        or was explicitly assigned.
      */
      if (!(field->charset()->state & MY_CS_PRIMARY) ||
          column_has_explicit_collation ||
          (field->charset() == &my_charset_utf8mb4_0900_ai_ci &&
           share->table_charset != &my_charset_utf8mb4_0900_ai_ci)) {
        packet->append(STRING_WITH_LEN(" COLLATE "));
        packet->append(field->charset()->name);
      }
    }
```

**Comment by xiekeyi98** [comment]
@AndrewDi Thanks for your help.
Would you mind review my code again?

File: executor/show.go [hunk 1]
```diff
@@ -590,13 +590,29 @@ func (e *ShowExec) fetchShowCreateTable() error {
...
+
 	fmt.Fprintf(&buf, "CREATE TABLE %s (\n", escape(tb.Meta().Name, sqlMode))
 	var pkCol *table.Column
 	var hasAutoIncID bool
 	for i, col := range tb.Cols() {
 		fmt.Fprintf(&buf, "  %s %s", escape(col.Name, sqlMode), col.GetTypeDesc())
 		if col.Charset != "binary" {
-			fmt.Fprintf(&buf, " CHARSET %s COLLATE %s", col.Charset, col.Collate)
+			if col.Charset != tblCharset || col.HasExplictedCollation() {
+				fmt.Fprintf(&buf, " CHARSET %s", col.Charset)
```
Comment (AndrewDi) [inline executor/show.go]:
```suggestion
				fmt.Fprintf(&buf, " CHARSET %s COLLATE %s",  col.Charset, col.Collate)
```

File: executor/show.go [hunk 2]
```diff
@@ -590,13 +590,29 @@ func (e *ShowExec) fetchShowCreateTable() error {
...
+			}
+			if col.Collate != tblCollate || col.HasExplictedCollation() {
+				fmt.Fprintf(&buf, " COLLATE %s", col.Collate)
+			}
```
Comment (AndrewDi) [inline executor/show.go]: I think following check is useless...
```golang
if col.Collate != tblCollate || col.HasExplictedCollation() {
				fmt.Fprintf(&buf, " COLLATE %s", col.Collate)
			}
```

File: executor/show.go (hunk 1 above)
Comment (xiekeyi98) [reply executor/show.go]: I think now what I write is more like what MYSQL writes, how do you think it ?

File: executor/show.go (hunk 2 above)
Comment (xiekeyi98) [reply executor/show.go]: I just want to be more like MYSQL...

Comment (AndrewDi) [reply executor/show.go]: ok

**Comment by xiekeyi98** [comment]
Because I don't understand how MYSQL handles it,  I rolled back some commits.
This PR will only resolve table && column are inconsistent.
I will create new issues to explain more.

**Pull request merged**